### DISCUSSION TITLE
Question about returnMapping functionality

### DISCUSSION THREAD

**Discussion by TheGreatCid** [idaholab/moose#20297] [Q&A General] [answered]
I have a question about the [`returnMappingSolve`](https://mooseframework.inl.gov/docs/doxygen/modules/classSingleVariableReturnMappingSolution.html#a83a7781c826b58ac3473e8f9cd00ad45) class.

I have a set of residual calculations that will return undefined values when a negative value is introduced by the return mapping solve. Obviously, this breaks the solver and leads to the simulation failing. 
Is there any way to prevent the return mapping from suggesting negative values while not harming the convergence of the return mapping?

**Comment by GiudGiud** [comment, marked as answer]
@jiangwen84 @hugary1995 any of you know about this particular system? Or knows who knows?

**Reply by hugary1995** [reply]
I discussed with @TheGreatCid offline about this. I'm posting here what we discovered in case this is helpful for anyone in the future:

In his case, the yield surface is undefined with negative plastic strain rates. The best way to handle this is to let the return mapping algorithm respect the lower bound of zero. There are several ways to achieve this, e.g. line search, trust region, bisection methods, etc. The `SingleVariableReturnMappingSolution` class in moose offers an interface to provide lower and upper bounds. He simply needs to set `_check_range = true;` in the constructor and override the method `minimumPermissibleValue`:
```cpp
virtual Real minimumPermissibleValue(const Real &) const override { return 0; }
```
Then the return mapping will make sure values passed into the `computeResidual` and `computeDerivative` methods always respect the lower bound.

**Reply by jiangwen84** [reply]
@dschwen is working on adding line search to the return mapping. I believe he also mentioned about setting bounds.
